In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from scipy import stats 
import matplotlib.pyplot as plt

In [14]:
Reactions = pd.read_csv("Reactions.csv",index_col=0)

In [15]:
Reactions.head()

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2                                   NaN   
1  97522e57-d9ab-4bd6-97bf-c24d952602d2  5d454588-283d-459d-915d-c48a2cb4c27f   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2  92b87fa5-f271-43e0-af66-84fac21052e6   
3  97522e57-d9ab-4bd6-97bf-c24d952602d2  163daa38-8b77-48c9-9af6-37a6c1447ac2   
4  97522e57-d9ab-4bd6-97bf-c24d952602d2  34e8add9-0206-47fd-a501-037b994650a2   

      Type             Datetime  
0      NaN  2021-04-22 15:17:15  
1  disgust  2020-11-07 09:43:50  
2  dislike  2021-06-17 12:22:51  
3   scared  2021-04-18 05:13:58  
4  disgust  2021-01-06 19:13:01

In [16]:
Reactions.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25553 entries, 0 to 25552
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Content ID  25553 non-null  object
 1   User ID     22534 non-null  object
 2   Type        24573 non-null  object
 3   Datetime    25553 non-null  object
dtypes: object(4)
memory usage: 998.2+ KB


In [17]:
Reactions.describe()

Content ID  \
count                                  25553   
unique                                   980   
top     4b2d0fff-3b4f-43ca-a7df-c430479cb9ba   
freq                                      49   

                                     User ID   Type             Datetime  
count                                  22534  24573                25553  
unique                                   500     16                25542  
top     c76c3393-88e2-47b0-ac37-dc4f2053f5a5  heart  2020-10-29 20:51:08  
freq                                      65   1622                    2

In [18]:
Reactions.Datetime

0        2021-04-22 15:17:15
1        2020-11-07 09:43:50
2        2021-06-17 12:22:51
3        2021-04-18 05:13:58
4        2021-01-06 19:13:01
                ...         
25548    2020-06-27 09:46:48
25549    2021-02-16 17:17:02
25550    2020-09-12 03:54:58
25551    2020-11-04 20:08:31
25552    2021-01-04 04:55:11
Name: Datetime, Length: 25553, dtype: object

In [19]:
Reactions.Type.unique()

array([nan, 'disgust', 'dislike', 'scared', 'interested', 'peeking',
       'cherish', 'hate', 'love', 'indifferent', 'super love',
       'intrigued', 'worried', 'like', 'heart', 'want', 'adore'],
      dtype=object)

In [20]:
Reactions.Type.value_counts()

Type
heart          1622
scared         1572
peeking        1559
hate           1552
interested     1549
dislike        1548
adore          1548
want           1539
love           1534
disgust        1526
like           1520
super love     1519
indifferent    1512
cherish        1501
worried        1497
intrigued      1475
Name: count, dtype: int64

We can drop the User ID column. We'll also need to drop rows with missing values, and change the "Datetime" column to a datetime data type. Finally, we'll need to change the "Type" column name like we did for the Content table.

In [22]:
# drop user ID column
Reactions.drop(labels = ['User ID'],axis=1,inplace=True)

In [24]:
# convert dates to datetime
Reactions['Datetime'] = pd.to_datetime(Reactions['Datetime'])

In [25]:
#drop rows with missing values
Reactions.dropna(inplace=True)

In [26]:
#rename column
Reactions.rename(columns={"Type":'Reaction_Type'},inplace=True)

In [27]:
Reactions.info()

<class 'pandas.core.frame.DataFrame'>
Index: 24573 entries, 1 to 25552
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Content ID     24573 non-null  object        
 1   Reaction_Type  24573 non-null  object        
 2   Datetime       24573 non-null  datetime64[ns]
dtypes: datetime64[ns](1), object(2)
memory usage: 767.9+ KB


In [28]:
Reactions.head()

Content ID Reaction_Type            Datetime
1  97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2020-11-07 09:43:50
2  97522e57-d9ab-4bd6-97bf-c24d952602d2       dislike 2021-06-17 12:22:51
3  97522e57-d9ab-4bd6-97bf-c24d952602d2        scared 2021-04-18 05:13:58
4  97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-01-06 19:13:01
5  97522e57-d9ab-4bd6-97bf-c24d952602d2    interested 2020-08-23 12:25:58

In [30]:
#import second Dataset
Reactiontypes = pd.read_csv('ReactionTypes.csv',index_col=0)


In [31]:
Reactiontypes.head()

Type Sentiment  Score
0       heart  positive     60
1        want  positive     70
2     disgust  negative      0
3        hate  negative      5
4  interested  positive     30

In [32]:
Reactiontypes.describe()

Score
count  16.000000
mean   39.625000
std    26.901983
min     0.000000
25%    14.250000
50%    40.000000
75%    66.250000
max    75.000000

In [33]:
#rename type column
Reactiontypes.rename(columns={'Type':'Reaction_Type'},inplace=True)
Reactiontypes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16 entries, 0 to 15
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Reaction_Type  16 non-null     object
 1   Sentiment      16 non-null     object
 2   Score          16 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 512.0+ bytes


In [34]:
Reactiontypes.value_counts()


Reaction_Type  Sentiment  Score
adore          positive   72       1
cherish        positive   70       1
disgust        negative   0        1
dislike        negative   10       1
hate           negative   5        1
heart          positive   60       1
indifferent    neutral    20       1
interested     positive   30       1
intrigued      positive   45       1
like           positive   50       1
love           positive   65       1
peeking        neutral    35       1
scared         negative   15       1
super love     positive   75       1
want           positive   70       1
worried        negative   12       1
Name: count, dtype: int64

In [35]:
#import data from content database
Content = pd.read_csv('Content.csv',index_col=0)

In [36]:
Content.head()

Content ID                               User ID  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2  8d3cd87d-8a31-4935-9a4f-b319bfe05f31   
1  9f737e0a-3cdd-4d29-9d24-753f4e3be810  beb1f34e-7870-46d6-9fc7-2e12eb83ce43   
2  230c4e4d-70c3-461d-b42c-ec09396efb3f  a5c65404-5894-4b87-82f2-d787cbee86b4   
3  356fff80-da4d-4785-9f43-bc1261031dc6  9fb4ce88-fac1-406c-8544-1a899cee7aaf   
4  01ab84dd-6364-4236-abbb-3f237db77180  e206e31b-5f85-4964-b6ea-d7ee5324def1   

    Type        Category                                                URL  
0  photo        Studying  https://socialbuzz.cdn.com/content/storage/975...  
1  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/9f7...  
2  photo  healthy eating  https://socialbuzz.cdn.com/content/storage/230...  
3  photo      technology  https://socialbuzz.cdn.com/content/storage/356...  
4  video            food  https://socialbuzz.cdn.com/content/storage/01a...

In [37]:
Content.describe()

Content ID  \
count                                   1000   
unique                                  1000   
top     97522e57-d9ab-4bd6-97bf-c24d952602d2   
freq                                       1   

                                     User ID   Type    Category  \
count                                   1000   1000        1000   
unique                                   446      4          41   
top     72d2587e-8fae-4626-a73d-352e6465ba0f  photo  technology   
freq                                       8    261          71   

                                                      URL  
count                                                 801  
unique                                                801  
top     https://socialbuzz.cdn.com/content/storage/975...  
freq                                                    1

In [39]:
Content.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Content ID  1000 non-null   object
 1   User ID     1000 non-null   object
 2   Type        1000 non-null   object
 3   Category    1000 non-null   object
 4   URL         801 non-null    object
dtypes: object(5)
memory usage: 46.9+ KB


In [41]:
Content.value_counts()

Content ID                            User ID                               Type   Category         URL                                                                            
004e820e-49c3-4ba2-9d02-62db0065410c  9a5aca53-6263-485a-bf77-e60c4608fd26  audio  tennis           https://socialbuzz.cdn.com/content/storage/004e820e-49c3-4ba2-9d02-62db0065410c    1
a784c1dd-e2ab-408d-a0c3-226a51d420e1  f86853a2-acda-4319-a50c-03f014248042  audio  technology       https://socialbuzz.cdn.com/content/storage/a784c1dd-e2ab-408d-a0c3-226a51d420e1    1
a429d355-8983-40a4-aa2b-c3f47acd5499  e8c9386c-8fbc-40cc-9587-ccb912ec5c10  photo  veganism         https://socialbuzz.cdn.com/content/storage/a429d355-8983-40a4-aa2b-c3f47acd5499    1
a4f6b4c2-36bd-4232-bef8-66577a601163  1ee25417-3648-4220-9b8f-fc82d1ef450c  photo  animals          https://socialbuzz.cdn.com/content/storage/a4f6b4c2-36bd-4232-bef8-66577a601163    1
a52f38d8-1c2e-4adc-b451-01c49f650195  afab7336-04d3-45fe-8917-0327c56e7435  GIF 

We won't need the User ID or URL columns for our analysis, so let's drop those. Then let's check the values in the category column to make sure they make sense given our data.

In [42]:

Content.drop(labels=['User ID','URL'],axis=1,inplace=True)


In [43]:
#Check values in category columns
Content['Category'].value_counts()

Category
technology           71
animals              67
travel               67
culture              63
science              63
fitness              61
food                 61
healthy eating       61
cooking              60
soccer               58
tennis               58
education            57
dogs                 56
studying             55
veganism             48
public speaking      48
Fitness               5
Animals               4
Science               4
"soccer"              3
"culture"             3
Soccer                3
"dogs"                2
Education             2
Studying              2
Travel                2
Food                  2
"veganism"            1
"public speaking"     1
Public Speaking       1
"technology"          1
"cooking"             1
Healthy Eating        1
"studying"            1
"food"                1
Culture               1
"tennis"              1
Technology            1
"animals"             1
Veganism              1
"science"             1
Name: c

We'll need to clean up these categories by removing punctuation and capitalization.


In [44]:
#strip quote mark and lowercase all values
Content['Category'] = Content['Category'].apply(lambda x:x.strip('"').lower())

In [45]:
Content.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Content ID  1000 non-null   object
 1   Type        1000 non-null   object
 2   Category    1000 non-null   object
dtypes: object(3)
memory usage: 31.2+ KB


In [46]:
#checking if all categories are in lower case
Content['Category'].value_counts()

Category
technology         73
animals            72
travel             69
science            68
culture            67
fitness            66
food               64
soccer             64
healthy eating     62
cooking            61
tennis             59
education          59
studying           58
dogs               58
public speaking    50
veganism           50
Name: count, dtype: int64

In [47]:
#rename content to content_type
Content.rename(columns={'Type':'Content_Type'},inplace=True)
Content['Content_Type'].value_counts()
     

Content_Type
photo    261
video    259
GIF      244
audio    236
Name: count, dtype: int64

Our data is clean, based on the client's specifications (remove rows with missing values, drop unnecessary colunmns, and ensure data types match the column values).

## Creating a joined table for analysis (Data Modelling)

In [48]:
#merge tables on content ID columns
top_performers = Reactions.merge(Content,on='Content ID')
top_performers

Content ID Reaction_Type            Datetime  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2020-11-07 09:43:50   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       dislike 2021-06-17 12:22:51   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2        scared 2021-04-18 05:13:58   
3      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-01-06 19:13:01   
4      97522e57-d9ab-4bd6-97bf-c24d952602d2    interested 2020-08-23 12:25:58   
...                                     ...           ...                 ...   
24568  75d6b589-7fae-4a6d-b0d0-752845150e56       dislike 2020-06-27 09:46:48   
24569  75d6b589-7fae-4a6d-b0d0-752845150e56     intrigued 2021-02-16 17:17:02   
24570  75d6b589-7fae-4a6d-b0d0-752845150e56    interested 2020-09-12 03:54:58   
24571  75d6b589-7fae-4a6d-b0d0-752845150e56       worried 2020-11-04 20:08:31   
24572  75d6b589-7fae-4a6d-b0d0-752845150e56       cherish 2021-01-04 04:55:11   

      Content_Type    Category  
0            photo    studying  
1            photo    studying  
2            photo    studying  
3            photo    studying  
4            photo    studying  
...            ...         ...  
24568        audio  technology  
24569        audio  technology  
24570        audio  technology  
24571        audio  technology  
24572        audio  technology  

[24573 rows x 5 columns]

In [49]:
#merge tables on reaction type column
top_performers = top_performers.merge(Reactiontypes,on='Reaction_Type')
top_performers

Content ID Reaction_Type            Datetime  \
0      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2020-11-07 09:43:50   
1      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-01-06 19:13:01   
2      97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust 2021-04-09 02:46:20   
3      9f737e0a-3cdd-4d29-9d24-753f4e3be810       disgust 2021-03-28 21:15:26   
4      230c4e4d-70c3-461d-b42c-ec09396efb3f       disgust 2020-08-04 05:40:33   
...                                     ...           ...                 ...   
24568  435007a5-6261-4d8b-b0a4-55fdc189754b         adore 2020-10-04 22:26:33   
24569  435007a5-6261-4d8b-b0a4-55fdc189754b         adore 2020-09-18 10:50:50   
24570  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore 2020-10-31 03:58:44   
24571  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore 2020-06-25 15:12:29   
24572  4e4c9690-c013-4ee7-9e66-943d8cbd27b7         adore 2020-12-17 16:32:57   

      Content_Type        Category Sentiment  Score  
0            photo        studying  negative      0  
1            photo        studying  negative      0  
2            photo        studying  negative      0  
3            photo  healthy eating  negative      0  
4            photo  healthy eating  negative      0  
...            ...             ...       ...    ...  
24568        audio        veganism  positive     72  
24569        audio        veganism  positive     72  
24570          GIF         culture  positive     72  
24571          GIF         culture  positive     72  
24572          GIF         culture  positive     72  

[24573 rows x 7 columns]

We have over 24,573 rows of data. Let's use this table to answer our business question regarding the top five categories with the largest aggregate popularity.
The measure of popularity provided to us by the client is the "Score" column. To find the highest aggregate popularity for each category, we simply have to sum this value by category.

## Top Category Analysis

In [52]:
## Have to drop Datetime , datetime64 type does not support sum operations

top_performers.drop(labels=['Datetime'],axis=1,inplace=True)


In [53]:
top_performers.head()

Content ID Reaction_Type Content_Type  \
0  97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust        photo   
1  97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust        photo   
2  97522e57-d9ab-4bd6-97bf-c24d952602d2       disgust        photo   
3  9f737e0a-3cdd-4d29-9d24-753f4e3be810       disgust        photo   
4  230c4e4d-70c3-461d-b42c-ec09396efb3f       disgust        photo   

         Category Sentiment  Score  
0        studying  negative      0  
1        studying  negative      0  
2        studying  negative      0  
3  healthy eating  negative      0  
4  healthy eating  negative      0

In [58]:
#group by category and calculate total score
top_five = top_performers.groupby('Category').sum().sort_values(by='Score',ascending=False)


In [59]:
top_five.head(5)

Content ID  \
Category                                                            
animals         4fa14453-7b29-4302-b51f-9aa23b472c1b4478d98e-4...   
science         5118e9c5-1377-4cc5-a486-65b35b7b7b7634a3747a-0...   
healthy eating  9f737e0a-3cdd-4d29-9d24-753f4e3be810230c4e4d-7...   
technology      49367a0f-1fd5-4417-a7c9-4a154e1ae0ad49367a0f-1...   
food            81abd65a-3b76-4574-a0a7-db6bf7184ae281abd65a-3...   

                                                    Reaction_Type  \
Category                                                            
animals         disgustdisgustdisgustdisgustdisgustdisgustdisg...   
science         disgustdisgustdisgustdisgustdisgustdisgustdisg...   
healthy eating  disgustdisgustdisgustdisgustdisgustdisgustdisg...   
technology      disgustdisgustdisgustdisgustdisgustdisgustdisg...   
food            disgustdisgustdisgustdisgustdisgustdisgustdisg...   

                                                     Content_Type  \
Category                                                            
animals         audioaudiophotophotophotophotophotovideovideov...   
science         GIFGIFGIFvideovideovideovideovideoaudioaudioau...   
healthy eating  photophotophotoGIFGIFGIFGIFGIFGIFGIFGIFGIFGIFv...   
technology      photophotovideovideovideoGIFaudioaudioaudioaud...   
food            videovideovideovideovideovideovideovideovideoa...   

                                                        Sentiment  Score  
Category                                                                  
animals         negativenegativenegativenegativenegativenegati...  74965  
science         negativenegativenegativenegativenegativenegati...  71168  
healthy eating  negativenegativenegativenegativenegativenegati...  69339  
technology      negativenegativenegativenegativenegativenegati...  68738  
food            negativenegativenegativenegativenegativenegati...  66676

In [60]:
top_five.Score

Category
animals            74965
science            71168
healthy eating     69339
technology         68738
food               66676
culture            66579
travel             64880
cooking            64756
soccer             57783
education          57436
fitness            55323
studying           54269
dogs               52511
tennis             50339
veganism           49619
public speaking    49264
Name: Score, dtype: int64

There's our answer - the top five categories by aggregate popularity are animals, science, healthy eating, technology, and food. We can export this cleaned dataset to a CSV to share with the client.

In [63]:
top_five['Score'].to_frame().head(5)

Score
Category             
animals         74965
science         71168
healthy eating  69339
technology      68738
food            66676

In [64]:
top_five_df = top_five['Score'].to_frame().head(5) 

In [65]:
top_five_df

Score
Category             
animals         74965
science         71168
healthy eating  69339
technology      68738
food            66676

In [66]:
top_five_df.to_csv('client_data.csv')


In [67]:
pd.read_csv("client_data.csv",index_col=0)

Score
Category             
animals         74965
science         71168
healthy eating  69339
technology      68738
food            66676